### PR TITLE
[SPARK-21721][SQL] Clear FileSystem deleteOnExit cache when paths are successfully removed

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

We put staging path to delete into the deleteOnExit cache of `FileSystem` in case of the path can't be successfully removed. But when we successfully remove the path, we don't remove it from the cache. We should do it to avoid continuing grow the cache size.

## How was this patch tested?

Added a test.

### CODE REVIEW

**Comment by viirya** [comment]
There is a test which doesn't drop the table it created...

**Comment by viirya** [comment]
retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 1]
```diff
@@ -423,7 +423,14 @@ case class InsertIntoHiveTable(
     // Attempt to delete the staging directory and the inclusive files. If failed, the files are
     // expected to be dropped at the normal termination of VM since deleteOnExit is used.
     try {
-      createdTempDir.foreach { path => path.getFileSystem(hadoopConf).delete(path, true) }
+      createdTempDir.foreach { path =>
+        path.getFileSystem(hadoopConf).delete(path, true)
+        // If we successfully delete the staging directory, remove it from FileSystem's cache.
+        val fs = path.getFileSystem(hadoopConf)
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: You call getFileSystem twice; you can reuse it.
I suppose this adds an additional check to see if the file exists, but this won't be called that often.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 2]
```diff
@@ -156,69 +156,71 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
-      Order(8, "Swift", "Hybrid", 485, "2015-02-19", "John S", "Redwood City", "CA", 20152),
-      Order(9, "Atlas", "Split", 303, "2015-02-28", "Jones S", "San Mateo", "CA", 20152))
-
-    val orderUpdates = Seq(
-      Order(1, "Atlas", "MTB", 434, "2015-01-07", "John D", "Pacifica", "CA", 20151),
-      Order(11, "Swift", "YFlikr", 137, "2015-01-23", "John D", "Hayward", "CA", 20151))
-
-    orders.toDF.createOrReplaceTempView("orders1")
-    orderUpdates.toDF.createOrReplaceTempView("orderupdates1")
+    withTable("orders", "orderupdates") {
```
Comment (srowen) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Is this intended to be part of this PR?

Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: The newly added test has been bitten by an existing test which doesn't drop created table. Those old tests all the same. I can move them to another minor PR if you think it's necessary.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala (hunk 1 above)
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: Oh. right. Thanks.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 3]
```diff
@@ -2021,4 +2038,22 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
       checkAnswer(table.filter($"p" === "p1\" and q=\"q1").select($"a"), Row(4))
     }
   }
+
+  test("SPARK-21721: Clear FileSystem deleterOnExit cache if path is successfully removed") {
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: For reviewers: This is the only one test added. Above changes to old tests are going to drop the created tables in those old tests.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 4]
```diff
@@ -423,7 +423,14 @@ case class InsertIntoHiveTable(
     // Attempt to delete the staging directory and the inclusive files. If failed, the files are
     // expected to be dropped at the normal termination of VM since deleteOnExit is used.
     try {
-      createdTempDir.foreach { path => path.getFileSystem(hadoopConf).delete(path, true) }
+      createdTempDir.foreach { path =>
+        val fs = path.getFileSystem(hadoopConf)
+        fs.delete(path, true)
```
Comment (dbolshak) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: `FileSystem.delete returns true if delete is successful else false.` So it makes sense to re-write a bit
```
if (fs.delete(path, true)) {
fs.cancelDeleteOnExit(path)
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Usually, each test case should be independent. Could you isolate your new test case instead? It would be better when we backport your PR later.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Yes. Please revert all the unrelated changes.

Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: OK.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala (hunk 4 above)
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: Great. Thanks.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.2

**Comment by gatorsmile** [comment]
Tried to merge to Spark 2.1, but I hit a conflict. Thus, please let me know if anybody wants it in 2.1. Thanks!

**Comment by yzheng616** [comment]
Please try to fix it in 2.1 too. We have a product running on this version Spark. Thanks a lot!

**Comment by gatorsmile** [comment]
cc @viirya

**Comment by viirya** [comment]
@yzheng616 @gatorsmile Ok. I will backport it to 2.1.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 5]
```diff
@@ -2021,4 +2021,22 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+      deleteOnExitField.setAccessible(true)
+
+      val fs = FileSystem.get(spark.sparkContext.hadoopConfiguration)
+      val setOfPath = deleteOnExitField.get(fs).asInstanceOf[Set[Path]]
+
+      val testData = sparkContext.parallelize(1 to 10).map(i => TestData(i, i.toString)).toDF()
+      sql("CREATE TABLE test21721 (key INT, value STRING)")
+      val pathSizeToDeleteOnExit = setOfPath.size()
+
+      (0 to 10).foreach(_ => testData.write.mode(SaveMode.Append).insertInto("test1"))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: please fix the typo, should be `test21721` instead of `test1`

Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Will fix now.
